test/integration/src/test/java/io/pravega/test/integration/RestoreBackUpDataRecoveryTest.java

```diff
@@ -423,8 +435,312 @@ public void testDurableDataLogFail() throws Exception {
+                "R" + RANDOM.nextInt(Integer.MAX_VALUE));
+        readAllEvents(STREAM2, clientFactory, readerGroupManager, "RG" + RANDOM.nextInt(Integer.MAX_VALUE),
+                "R" + RANDOM.nextInt(Integer.MAX_VALUE));
+        log.info("Read all events again to verify that segments were recovered.");
+    }
+
+    /**
+     * Tests the data recovery scenario with multiple segment containers. Segments recovery is attained using multiple
+     * debug segment containers as well.
+     *  What test does, step by step:
+     *  1. Starts Pravega locally with just 4 segment containers.
+     *  2. Writes 300 events to two different segments.
+     *  3. Waits for all segments created to be flushed to the long term storage.
+     *  4. Shuts down the controller, segment store and bookeeper/zookeeper.
+     *  5. Deletes container metadata segment and its attribute segment from the old LTS.
+     *  5. Starts 4 debug segment containers using a new bookeeper/zookeeper and the old LTS.
+     *  6. Re-creates the container metadata segment in Tier1 and let's it flushed to the LTS.
+     *  7. Starts segment store and controller.
+     *  8. Reads all 600 events again.
+     * @throws Exception    In case of an exception occurred while execution.
+     */
+    @Test(timeout = 180000)
+    public void testDurableDataLogFailRecoveryMultipleContainers() throws Exception {
+        int instanceId = 0;
+        int containerCount = 4;
+        // Creating a long term storage only once here.
+        this.storageFactory = new InMemoryStorageFactory(executorService());
+        log.info("Created a long term storage.");
+
+        // Start a new BK & ZK, segment store and controller
+        this.bookKeeperStarter = setUpNewBK(instanceId++);
+        @Cleanup
+        SegmentStoreStarter segmentStoreStarter = startSegmentStore(this.storageFactory, null, containerCount);
+        @Cleanup
+        ControllerStarter controllerStarter = startController(this.bookKeeperStarter.bkPort, segmentStoreStarter.servicePort,
+                containerCount);
+
+        // Create two streams for writing data onto two different segments
+        createScopeStream(controllerStarter.controller, SCOPE, STREAM1);
+        createScopeStream(controllerStarter.controller, SCOPE, STREAM2);
+        log.info("Created two streams.");
+
+        @Cleanup
+        ConnectionFactory connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder()
+                .controllerURI(controllerStarter.controllerURI).build());
+        @Cleanup
+        ClientFactoryImpl clientFactory = new ClientFactoryImpl(SCOPE, controllerStarter.controller, connectionFactory);
+        @Cleanup
+        ReaderGroupManager readerGroupManager = new ReaderGroupManagerImpl(SCOPE, controllerStarter.controller, clientFactory);
+
+        log.info("Writing events on to stream: {}", STREAM1);
+        writeEvents(STREAM1, clientFactory); // write 300 events on one segment
```

Comment:
Instead of duplicating code , please extract out method that takes two parameters - number of containers and boolean for whether or not to use transactions.

Generally try to follow the DRY (Don't Repeat Yourself)  principle by extracting common functionality in smaller methods,

Reply:
Done.